### PR TITLE
Move realm data migrations inside the `RealmContextFactory`

### PR DESCRIPTION
**Pull request by peppy**
Feels like a better location. Can also use the helper method to centralise the ruleset lookup logic. Will become more important as we begin to migrate more data across.

I've tested this manually with an empty and non-empty realm database (and ensured that settings still do migrate as expected).

- [x] Depends on https://github.com/ppy/osu/pull/15741.

### CODE REVIEW

File: osu.Game/Database/RealmContextFactory.cs [hunk 1]
```diff
@@ -96,6 +96,9 @@ public RealmContextFactory(Storage storage, string filename, IDatabaseContextFac
                 Filename += realm_extension;
 
             cleanupPendingDeletions();
+
+            // run after the above operation to ensure realm is already in a good (fully migrated) state.
```
Comment (bdach) [inline osu.Game/Database/RealmContextFactory.cs:100]: Confused by this comment, especially the mention of "fully migrated". Did this mean to say "fully cleaned up"?

File: osu.Game/Database/RealmContextFactory.cs [hunk 2]
```diff
@@ -163,6 +166,54 @@ private RealmConfiguration getConfiguration()
...
+                            string? shortName = getRulesetShortNameFromLegacyID(dkb.RulesetID.Value);
+
+                            if (string.IsNullOrEmpty(shortName))
+                                continue;
+
+                            realm.Add(new RealmRulesetSetting
+                            {
+                                Key = dkb.Key,
+                                Value = dkb.StringValue,
+                                // important: this RulesetStore must be the EF one.
```
Comment (bdach) [inline osu.Game/Database/RealmContextFactory.cs:203]: Should probably trim this comment, kind of a non sequitur here.

**Review by bdach — COMMENTED** [review]
Generally looks ok, save a few comment nits and the prereq

File: osu.Game/Database/RealmContextFactory.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Database/RealmContextFactory.cs:100]: I meant the realm level migrations, which are separate from data migrations (see `onMigration` stuff).

Comment (peppy) [reply osu.Game/Database/RealmContextFactory.cs]: I've rewritten the comments around this code to hopefully be easier to follow.

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**